cql3/restrictions/statement_restrictions.cc

```diff
@@ -1458,8 +1447,180 @@ bool is_satisfied_by(
+            return false;
+        } else if (!this->_value) { // Now that we're sure _value is valid, check it for null.
+            return true;
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+bound_t tighter_ub(const bound_t& a, const bound_t& b) {
+    return a.is_tighter_ub_than(b) ? a : b;
+}
+
+bound_t tighter_lb(const bound_t& a, const bound_t& b) {
+    return a.is_tighter_lb_than(b) ? a : b;
+}
+
+/// The type of the first child, if it exists.  Otherwise, byte_type.
+const abstract_type* child_type(const children_t& children) {
+    return children.empty() ? byte_type.get() : std::visit(overloaded_functor{
+            [] (bool) { return byte_type.get(); },
+            [] (const conjunction& c) { return child_type(c.children); },
+            [] (const binary_operator& op) {
+                return std::visit(overloaded_functor{
+                        [] (const token&) { return long_type.get(); },
+                        [] (const std::vector<column_value> cvs) {
+                            if (cvs.empty() || cvs.size() > 1) {
+                                throw std::logic_error(
+                                        format("unexpected conjunct with {} columns (should be a single column)",
+                                               cvs.size()));
+                            } else {
+                                return comparator(cvs[0]);
+                            }
+                        },
+                    }, op.lhs);
+            },
+        }, children[0]);
+}
+
+/// True iff op means bnd type of bound.
+bool matches(const operator_type* op, statements::bound bnd) {
+    static const std::vector<std::vector<const operator_type*>> operators{
+        {&operator_type::EQ, &operator_type::GT, &operator_type::GTE}, // These mean a lower bound.
+        {&operator_type::EQ, &operator_type::LT, &operator_type::LTE}, // These mean an upper bound.
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+bound_t get_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    return std::visit(overloaded_functor{
+            [&] (bool v) {
+                if (v) {
+                    return bound_t(byte_type);
+                } else {
+                    throw std::logic_error("empty set has no bounds");
+                }
+            },
+            [&] (const conjunction& conj) {
+                auto invoke_get_bound = [&] (const expression& e) { return get_bound(e, options, bnd); };
+                auto pick_tighter = is_start(bnd) ? &tighter_lb : &tighter_ub;
+                // All conjunction elements have the same LHS; this is how
+                // single_column_restrictions::add_restriction constructs it.  We pick the tightest among all
```

Comment:
Ok, I think I understand why the comparator function peeled the reversed_type - reversing changes the sort order but shouldn't change the meaning of <= etc.

I think this should be made visible in the function's name.

Reply:
Renamed to `get_value_comparator`.